src/core/llm/services/vercel.ts

```diff
@@ -92,14 +95,53 @@ export class VercelLLMService implements ILLMService {
             if (tool) {
                 acc[toolName] = {
                     inputSchema: jsonSchema(tool.parameters),
-                    execute: async (args: unknown) => {
+                    execute: async (args: unknown, _options?: { toolCallId?: string }) => {
                         try {
-                            return await this.toolManager.executeTool(
+                            const rawResult = await this.toolManager.executeTool(
                                 toolName,
                                 args as Record<string, unknown>,
                                 this.sessionId
                             );
+
+                            // Queue raw, unfiltered result under the specific toolCallId
+                            const callId = _options?.toolCallId ?? this.sessionId;
+                            const queue = this.rawResultsByCallId.get(callId) ?? [];
```

Comment:
don't fallback call id to session id, make the toolCallId non optional (vercel always provides it)

and use that

https://ai-sdk.dev/docs/reference/ai-sdk-core/stream-text